service/history/shard/controller_impl.go

```diff
@@ -75,6 +76,20 @@ type (
 		ownership            *ownership
 		status               int32
 		taggedMetricsHandler metrics.Handler
+		// shardCountSubscriptions is a list of subscriptions that receive shard count updates whenever the set of
+		// shards that this controller owns changes.
+		shardCountSubscriptions []*shardCountSubscription
+	}
+	// controllerShardCounter is a wrapper around ControllerImpl that implements the ShardCounter interface. We don't want
+	// to add Subscribe directly to ControllerImpl because it's not clear what you're subscribing to if you call Subscribe
+	// directly on the controller. Also, we don't want to export the Subscribe functionality of ControllerImpl.
```

Comment:
It still feels a bit weird to me that controllerShardCounter and Subscription composes ControllerImpl but need to access and change its internal fields/states.

If the idea is just to avoid confusion for exported methods on Controller, we can also define private/non-exported methods on Controller, let another struct embed Controller and export those methods with a different naming.

Or it the method is called SubscribeShardCount(), does it make sense to call it directly on the controller?

Reply:
Renamed and removed the wrapper.